### PR TITLE
[RFC] [Testcase Refactoring]Make distributed data parallel tests device-agnostic using torch.accelerator APIs

### PR DESCRIPTION
**Pull request by zhaiyaxing001**
## Summary

This PR refactors `test/distributed/test_data_parallel.py` to be device-agnostic by replacing hardcoded CUDA references (`TEST_MULTIGPU`, `@onlyCUDA`, `.cuda()`, `torch.cuda.device_count()`) with `torch.accelerator` APIs and `instantiate_device_type_tests`-injected `device` parameters.

## Motivation

The original test had hardcoded CUDA assumptions:
- `from torch.cuda.amp import autocast` and `TEST_CUDA`/`TEST_MULTIGPU` imports
- `@onlyCUDA` and `@skip_but_pass_in_sandcastle_if(not TEST_MULTIGPU, ...)` decorators
- Hardcoded `"cuda:0"`, `"cuda:1"`, `.cuda()`, `torch.cuda.device_count()` throughout
- Separate `test_scatter_cpu`/`test_scatter_gpu` and `test_gather_cpu`/`test_gather_gpu` methods
- No `device` parameter, no `hw_classification`

### 1. Import changes

```diff
-from torch.cuda.amp import autocast
-from torch.testing._internal.common_cuda import TEST_CUDA, TEST_MULTIGPU
+from torch.amp import autocast
 from torch.testing._internal.common_device_type import (
-    onlyCUDA,
+    onlyAccelerator,
+    skipCPUIf,
     ...
 )
 from torch.testing._internal.common_utils import (
+    TEST_MULTIACCELERATOR,
+    HardwareClassification,
     ...
 )
```

### 2. Add `hw_classification`

```python
class TestDataParallel(TestCase):
    hw_classification = HardwareClassification.ACCELERATOR

class TestDataParallelDeviceType(TestCase):
    hw_classification = HardwareClassification.ACCELERATOR
```

### 3. Replace device decorators (all test methods)

```diff
-@skip_but_pass_in_sandcastle_if(not TEST_MULTIGPU, "multi-GPU not supported")
+@onlyAccelerator
+@skip_but_pass_in_sandcastle_if(
+    not TEST_MULTIACCELERATOR, "multi-accelerator not supported"
+)
```

### 4. Add `device` parameter (24 methods)

All test method signatures get `device` parameter.

### 5. Replace hardcoded device references (~60 occurrences)

| Before | After |
|--------|-------|
| `device="cuda"` / `"cuda:0"` / `"cuda:1"` | `device=device` |
| `.cuda()` / `.cuda(0)` / `.cuda(1)` | `.to(device)` / `.to(torch.device(dev_type, 1))` |
| `torch.cuda.device_count()` | `torch.accelerator.device_count()` |
| `torch.cuda.device(dev_id[0])` | `torch.accelerator.device_index(dev_id[0])` |
| `torch.device(f"cuda:{device_ids[0]}")` | `torch.device(dev_type, device_ids[0])` |
| `autocast()` | `autocast(device_type=dev_type)` |
| `@torch.autocast(device_type="cuda")` | `@torch.autocast(device_type=torch.device(device).type)` |

### 6. Merge test methods

`test_scatter_cpu` + `test_scatter_gpu` → single `test_scatter(self, device)`

`test_gather_cpu` + `test_gather_gpu` → single `test_gather(self, device)` with CPU gather test included

### 7. Conditional cudnn flags

```python
cudnn_ctx = (
    torch.backends.cudnn.flags(enabled=True, deterministic=True, benchmark=False)
    if self.device_type == "cuda"
    else contextlib.nullcontext()
)
```

### 8. Add `instantiate_device_type_tests`

```python
instantiate_device_type_tests(TestDataParallelDeviceType, globals(), except_for="cpu")
instantiate_device_type_tests(TestDataParallel, globals())
```

## Testing

| Environment | Result |
|-------------|--------|
| CPU | 42 tests skipped (no multi-accelerator) |
| NPU | 33 tests passed, 19 tests skipped (NPU hardware limitations) |

## Compatibility

- Uses `torch.accelerator.current_accelerator()` public API
- Uses `torch.amp.autocast` public API
- CUDA behavior unchanged
- Supports NPU, HPU, XPU and other accelerators
- Note: Some tests are skipped on NPU due to actual hardware limitations, not code bugs

### CODE REVIEW

File: test/distributed/test_data_parallel.py [hunk 1]
```diff
@@ -51,19 +54,19 @@ def forward(self, x):
                 return x * self.t_rg + self.t_not_rg
 
         m = TestModule(
-            torch.randn(100, device="cuda", requires_grad=True, dtype=torch.double)
+            torch.randn(100, device=DEVICE_TYPE, requires_grad=True, dtype=torch.float)
         )
         self.assertTrue(m.t_rg.requires_grad)
 
         dpm = nn.DataParallel(m, [0, 1])
-        inp = torch.randn(2, 100, device="cuda", dtype=torch.double)
```
Comment (wjlFlyer) [inline test/distributed/test_data_parallel.py:59]: @zhaiyaxing001 Why change dtype ? I'm afraid it may affect precision of result.

File: test/distributed/test_data_parallel.py [hunk 2]
```diff
@@ -39,7 +42,7 @@
 
 
 class TestDataParallel(TestCase):
-    @skip_but_pass_in_sandcastle_if(not TEST_MULTIGPU, "multi-GPU not supported")
+    @skip_but_pass_in_sandcastle_if(not TEST_MULTIACCELERATOR or DEVICE_TYPE == "npu", "gradcheck with float32 not reliable on NPU")
```
Comment (wjlFlyer) [inline test/distributed/test_data_parallel.py]: @zhaiyaxing001 why skip this test case on `npu`? Does it not work on npu ? If so, i think it's better to use blacklist(yaml, or json) to filter this test case on npu.

File: test/distributed/test_data_parallel.py (hunk 1 above)
Comment (zhaiyaxing001) [reply test/distributed/test_data_parallel.py:59]: our assessment is accurate; no amendments are required for the specified section. The exception pertaining to the NPU scenario has been duly clarified within the label, and the entire text has undergone a comprehensive review and correction.

File: test/distributed/test_data_parallel.py (hunk 2 above)
Comment (zhaiyaxing001) [reply test/distributed/test_data_parallel.py]: The tag has been removed. An issue has been submitted to indicate that the NPU does not support this capability.issue:https://gitcode.com/Ascend/pytorch/issues/4232,
The blacklist will be added as required.

File: test/distributed/test_data_parallel.py [hunk 3]
```diff
@@ -31,6 +33,9 @@
 
 NO_NCCL = not hasattr(torch.distributed, "ProcessGroupNCCL")
 
+_accelerator = torch.accelerator.current_accelerator()
+DEVICE_TYPE = _accelerator.type if _accelerator else "cpu"
```
Comment (wjlFlyer) [inline test/distributed/test_data_parallel.py]: I think it unnecessary to define `DEVICE_TYPE ` here. After using `instantiate_device_type_tests` to instantiate the class, every test method will has `device` parameter.

Comment (zhaiyaxing001) [reply test/distributed/test_data_parallel.py]: Yes, the dynamic device injection rectification needs to be completed; it has been completed.

**Review by wjlFlyer — APPROVED** [review]

File: test/distributed/test_data_parallel.py [hunk 4]
```diff
@@ -124,13 +127,13 @@ def test_parallel_apply(self):
...
-        i1 = torch.randn(2, 10, device="cuda:0", dtype=torch.float)
-        i2 = torch.randn(2, 10, device="cuda:1", dtype=torch.float)
-        with autocast():
+    @skip_but_pass_in_sandcastle_if(not TEST_MULTIACCELERATOR, "multi-accelerator not supported")
+    def test_parallel_apply_autocast(self, device):
+        l1 = nn.Linear(10, 5).to(device, torch.float)
+        l2 = nn.Linear(10, 5).to(device, torch.float)
+        i1 = torch.randn(2, 10, device=device, dtype=torch.float)
+        i2 = torch.randn(2, 10, device=device, dtype=torch.float)
+        with autocast(device_type=device):
```
Comment (fffrog) [inline test/distributed/test_data_parallel.py:143]: Nice work

File: test/distributed/test_data_parallel.py [hunk 5]
```diff
@@ -503,21 +506,21 @@ def gradient_penalty(net, x):
...
         self.assertEqual(2, len(grads))
         self.assertEqual(
-            torch.tensor([[0.25, 0.25, 0.25, 0.25]], device="cuda:0"), grads[0]
+            torch.tensor([[0.25, 0.25, 0.25, 0.25]], device=device), grads[0]
         )
-        self.assertEqual(torch.tensor([0.0], device="cuda:0"), grads[1])
+        self.assertEqual(torch.tensor([0.0], device=device), grads[1])
 
-    def _test_scatter(self, tensor):
+    def _test_scatter(self, tensor, device):
```
Comment (fffrog) [inline test/distributed/test_data_parallel.py]: Why does we add device here?

File: test/distributed/test_data_parallel.py [hunk 6]
```diff
@@ -531,16 +534,16 @@ def _test_scatter(self, tensor):
         self.assertEqual(x.grad[2:], grad.clone().zero_())
         _assertGradAndGradgradChecks(self, lambda y: dp.scatter(y, (0, 1)), (x,))
 
-    @skip_but_pass_in_sandcastle_if(not TEST_MULTIGPU, "multi-GPU not supported")
-    def test_scatter_cpu(self):
-        self._test_scatter(torch.randn((4, 4), dtype=torch.double))
+    @skip_but_pass_in_sandcastle_if(not TEST_MULTIACCELERATOR, "multi-accelerator not supported")
+    def test_scatter_cpu(self, device):
```
Comment (fffrog) [inline test/distributed/test_data_parallel.py]: We can consolidate `test_scatter_cpu`  and `test_scatter_gpu` into a new unified testcase named `test_scatter` and can consider to remove `_test_scatter` as well.

And the new `test_scatter` will be runned in both `CPU` and `CUDA` side.

File: test/distributed/test_data_parallel.py [hunk 7]
```diff
@@ -724,55 +727,55 @@ def _test_gather(self, output_device):
...
         self.assertEqual(inputs[1].grad, grad[1])
         _assertGradAndGradgradChecks(
             self, lambda x, y: dp.gather((x, y), output_device), inputs
         )
 
-    @skip_but_pass_in_sandcastle_if(not TEST_MULTIGPU, "multi-GPU not supported")
-    def test_gather_cpu(self):
-        self._test_gather(-1)
+    @skip_but_pass_in_sandcastle_if(not TEST_MULTIACCELERATOR, "multi-accelerator not supported")
+    def test_gather_cpu(self, device):
```
Comment (fffrog) [inline test/distributed/test_data_parallel.py]: Hi @zhaiyaxing001 ， this one is same as the `test_scatter_cpu`, but this one should be added `onlyAccelerator` to limit the execution.

File: test/distributed/test_data_parallel.py [hunk 8]
```diff
@@ -1074,8 +1074,8 @@ def forward(self, input):
         self.assertEqual(out.get_device(), 0)
         self.assertEqual(out, expected_out, atol=dtype2prec_DONTUSE[dtype], rtol=0)
 
-
-instantiate_device_type_tests(TestDataParallelDeviceType, globals())
+instantiate_device_type_tests(TestDataParallelDeviceType, globals(), except_for="cpu")
```
Comment (fffrog) [inline test/distributed/test_data_parallel.py:1095]: Remove `except_for=cpu` please, because `test_scatter` should be runned in CPU ,and the rest should be runned in accelerator, but we do NOT need to add `onlyAcclerator` for them, because the `TEST_MULTIACCELERATOR` can limit that only acceleraotr can run those testcases.

**Review by fffrog — CHANGES_REQUESTED** [review]

File: test/distributed/test_data_parallel.py (hunk 8 above)
Comment (zhaiyaxing001) [reply test/distributed/test_data_parallel.py:1095]: test_scatter is in TestDataParallel.Already removed

File: test/distributed/test_data_parallel.py (hunk 5 above)
Comment (zhaiyaxing001) [reply test/distributed/test_data_parallel.py]: test cases have been merged.

File: test/distributed/test_data_parallel.py (hunk 6 above)
Comment (zhaiyaxing001) [reply test/distributed/test_data_parallel.py]: Corrected, test cases have been merged.

File: test/distributed/test_data_parallel.py (hunk 7 above)
Comment (zhaiyaxing001) [reply test/distributed/test_data_parallel.py]: Corrected, test cases have been merged.

File: test/distributed/test_data_parallel.py [hunk 9]
```diff
@@ -689,10 +708,15 @@ def forward(self, x_real, x_complex):
                     lambda msg: f"{msg}\nEpoch {epoch}: weights differ for {n1} after optimizer step",
                 )
 
-    def _test_gather(self, output_device):
+    @skip_but_pass_in_sandcastle_if(not TEST_MULTIACCELERATOR, "multi-accelerator not supported")
+    def test_gather(self, device):
+        if isinstance(device, str) and device == "cpu":
+            raise unittest.SkipTest("dp.gather does not support CPU output_device")
```
Comment (fffrog) [inline test/distributed/test_data_parallel.py]: It would be better to use skipCPUIf(True, "dp.gather does not support CPU output_device")

**Review by fffrog — DISMISSED** [review]
SGTM.

File: test/distributed/test_data_parallel.py (hunk 9 above)
Comment (zhaiyaxing001) [reply test/distributed/test_data_parallel.py]: Already corrected use skipCPUIf

**Comment by can-gaa-hou** [comment]
@claude review this pr with pr review skill
